### ISSUE TITLE
Main inside package.json is referencing to es6 modules 

### ISSUE THREAD

**Issue by isidrok** [material-components/material-components-web#929]
### What MDC-Web Version are you using?

This happens with all components which have different versions, for instance the TextField is in version 0.3.1

### What are the steps to reproduce the bug?

1. Create a project with `npm init`
2. Install the TextField component
```
npm i --save  @material/textfield
```
3. Configure rollup or webpack to bundle the project and set babel to exclude  'node_modules/**'
4. Somewhere in your project use the textfield:
```javascript
import * as textfield from '@material/textfield';
mdcComponent = new textfield.MDCTextfield(element)
```
5. Check that the bundled app has the TextField component written in es6
6. Go to `node_modules/@material/textfield`
7. Change "main" from "index.js" to "dist/mdc.textfield.min.js"
8. Create the bundle again, note that this time it will not find the module
9. Change "main" back to "index.js"
9. Stop excluding  'node_modules/**' in the babel configuration
10. Should work fine this time

### What is the expected behavior?

The "main" property inside the package.json should reference to the transpiled and bundled module instead of the es6 version so developers don't have to transpile it in order to avoid incompatibilities.
For example the TextField component should be pointing to dist/mdc.textfield.min.js instead of index.js.

There's actually another problem because the dist versions are exporting the utilities differently. The following snippet should work equally in the original and the minified/transpiled version:

```javascript
import * as textfield from '@material/textfield';
mdcComponent = new textfield.MDCTextfield(element)
```
but using the later i get the following error:
_'MDCTextfield' is not exported by 'node_modules\@material\textfield\dist\mdc.textfield.min.js'_

### What is the actual behavior?

The "main" is pointing to index.js which contains the es6 module, and if I don't transpile it some things won't work, for example I cannot use Uglify.

_(uglify plugin) Error transforming bundle with 'uglify' plugin: Unexpected token: name (MDCFoundation)_

That's because it finds MDCFoundation defined as a class.

### Any other information you believe would be useful?

There is a proposal to specify two entry points or main references inside package.json:
```json
...
"main": "dist/app.min.js",
"jsnext:main": "index.js",
...
```

For more information refer to this issue: https://github.com/jsforum/jsforum/issues/5

**Comment by Shyam-Chen** [comment]
```js
[...]
  "module": "lib/es/index.js",
  "main": "lib/cjs/index.js",
[...]
```

**Comment by brunoscopelliti** [comment]
I had this same problem several time.
When generate the build using webpack, I managed to work around configuring babel-loader so that it also transpiles files under `node_modules/@material`.

```
{
  test: /\.js$/,
  include: [
    appFolder,
    path.resolve(__dirname, "node_modules/@material")
  ],
  loader: "babel-loader"
}
```

But I still don't get very far, cause the same issue is currently beating me in my unit tests (where I don't use webpack).
I use `ava` for unit tests. ava default behaviour is of transpiling test files, but not the source files... to avoid issue with my own source file I configured ava so that it always include babel-register.

```

**Comment by Shyam-Chen** [comment]
Temporarily use this.

```js
import { __moduleExports as mdRipple } from '@material/ripple/dist/mdc.ripple';

[].forEach.call(
  document.querySelectorAll('.mdc-button'),
  ripple => mdRipple.MDCRipple.attachTo(ripple)
);
```

**Comment by brunoscopelliti** [comment]
That worked (almost)... since I was already transpiling my own sources, I had just to modify the import like this:
```js
import { Ripple } from '@material/ripple/dist/mdc.ripple';
```
Still, looking forward for a better solution, that allows to remove the ugly import.
Thanks @Shyam-Chen

**Comment by isidrok** [comment]
For me the only option that works at the moment is the one @Shyam-Chen  suggests, I made a list of possible imports and those are the results:

```javascript
import imports from '@material/textfield/dist/mdc.textfield';
// which is equivalent to import {default as imports } from '@material/textfield/dist/mdc.textfield';
imports = undefined

import { MDCTextfield } from '@material/textfield/dist/mdc.textfield';
// Non-existent export 'MDCTextField' is imported from...

import * as imports from '@material/textfield/dist/mdc.textfield';
imports = {
	default: undefined,
	__moduleExports: {
		MDCTextfield,
		MDCTextfieldFoundation
	}
}

import * as imports  from '@material/textfield';
imports = {
    MDCTextfield,
    MDCTextfieldFoundation
}
```

Don't know why is webpack placing the exports under __moduleExports or if that is the normal behaviour though.

**Comment by walexnelson** [comment]
I can't tell you how long I spent trying to figure out all of the UglifyJS errors I was getting because of this. I'm running my JS through the babel-loader in webpack and have it excluding `node_modules` like I usually do. It took me a while to realize that the @material modules need to be transpiled too if you import them like you would a typical module, at least if you want to import `material-components-web` rather than importing the individual transpiled @material modules from their respective `./dist` folders.

At the very least I think this needs to be documented unless I've been searching in all the wrong places.

I'm currently using a similar approach to @brunoscopelliti  and it seems to be working. UglifyJS isn't choking on ES6+ syntax found in my bundle at least.

**Comment by isidrok** [comment]
Well if someone is wondering how to solve this directly with Babel I use the include option (it cannot be used with exclude, everything not included will be excluded).

```js
babel({
    include: [
       'node_modules/@material/**',
       'src/**'
    ]
})
```

**Comment by ciromattia** [comment]
Being bound to Laravel Mix I can't follow the _babel/include_ way to include material-components-web into my js, after some tough headaches in the meantime I workarounded including the whole dists like

`import * as mdc from 'material-components-web/dist/material-components-web';`

Still following for a less hackish solution ;)

**Comment by tyronedougherty** [comment]
I'd just like to +1 my support recommending the change to the `main` in the `package.json`s, so that they point at the `dist` folder instead. At the moment to solve it I'm having to reference the `dist` folder directly, for example:

```
import { MDCDialogFoundation } from '@material/dialog/dist/mdc.dialog';
```

It's a pretty common pattern across just about every other repository you can find on NPM. I'm not sure why it's not the case here. Some examples can be found in [my duplicate issue](https://github.com/material-components/material-components-web/issues/1539).

**Comment by tbrisbane** [comment]
Also +1 on this.

By referencing ES6+ code in the `main` entry, you're effectively forcing your library consumers to have to either care about transpilation or go digging through the `dist` folder (as Tyrone mentioned above).

**Comment by chriscasola** [comment]
Is there any movement on this? I can open a PR if there is consensus on what should happen.

**Comment by isidrok** [comment]
In case someone is having problems with `jest` and `babel-jest`, exclude `@material` modules from transformation:
```json
{
  "transform": {
    "^.+\\.jsx$": "babel-jest",
    "^.+\\.js$": "babel-jest"
  },
  "transformIgnorePatterns": [
    "node_modules/(?!@material)"
  ],
} 
```

**Comment by brneto** [comment]
I did this:
```
"jest": {
    "setupTestFrameworkScriptFile": "<rootDir>/src/setupTests.js",
    "transformIgnorePatterns": ["node_modules/(?!@material)"]
  }
```
But I'm still facing the following error:
```
import MDCTopAppBarAdapter from './adapter';
^^^^^^

    SyntaxError: Unexpected token import

      3 | import { bindActionCreators } from 'redux';
      4 | import { connect } from 'react-redux';
    > 5 | import { MDCTopAppBar } from '@material/top-app-bar';
        | ^
      6 | import * as actionCreators from '../../redux/actions';
      7 | import * as routerThunks from '../../redux/thunks/routerThunks';
      8 |

      at ScriptTransformer._transformAndBuildScript (node_modules/jest-runtime/build/script_transformer.js:402:17)
      at Object.<anonymous> (src/client/components/commons/NavBar.js:5:1)
```

Any advice about how would I to fix that?

**Comment by blowery** [comment]
+1 to getting this fixed. The general convention these days is to have package.json's `main` property point to commonJS exported es5 (with an optional UMD wrapper). You could also add a `module` property that points to es5 code exported using ES6 exports (but _no_ es6+ code there). 

There's really no current spec for exposing untranspiled es6+ code in a standard way. There's some movement around a `jsmain:next` property, but how to have clients reliably transpile the code is up in the air. See https://babeljs.io/blog/2018/06/26/on-consuming-and-publishing-es2015+-packages

In the current state, folks that want to use the core modules have to set up transpilation to use them, which opens the door to version mismatches and configuration mismatches and bugs.

**Comment by kfranqueiro** [comment]
We've changed our packages' `main` properties to point to `dist` resources in #3245. This will be available in 0.41.0.

Thanks everyone for the input and sorry for the delay. (Also thanks @moog16 for starting the ball rolling and doing some testing for the PR.)
